UPGRADING.md

```diff
@@ -4,7 +4,264 @@ Upgrading Guide
+
+`Error` has been replaced by `Event`, which represents a JSON payload that will be sent to Bugsnag's API.
+A large number of new accessors have been added to the `Event` class to allow for greater customization of error reports in callbacks.
+
+Several existing methods have been renamed, a full list of which is shown below:
+
+| v4.x API  | v5.x API |
+| ------------- | ------------- |
+| `Error#addToTab`  | `Event#addMetadata(String, String, Object)` |
+| `Error#clearTab`  | `Event#clearMetadata(String)` |
+| `Error#getDeviceData`  | `Event#getDevice` |
+| `Error#getException`  | `Event#getOriginalError` |
+| `Error#getHandledState`  | `Event#isUnhandled` |
+| `Error#getMetaData`  | `Event#getMetadata(String)` |
+| `Error#setDeviceId`  | `Event#getDevice` |
+| `Error#setExceptionMessage`  | `Event#setErrors` |
+| `Error#setExceptionName`  | `Event#setErrors` |
+| `Error#setMetaData`  | `Event#addMetadata(String, String, Object)` |
+| `Error#setUserEmail`  | `Event#setUser(String, String, String)` |
+| `Error#setUserId`  | `Event#setUser(String, String, String)` |
+| `Error#setUserName`  | `Event#setUser(String, String, String)` |
+
+See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.
+
+## BeforeNotify/BeforeSend/Callback
+
+These three callbacks have been superseded and replaced by `OnError`, a single callback which runs immediately after an error has occurred. This can run globally on all errors, or on an individual error.
+
+```
+// run on all errors
+Bugsnag.addOnError { event ->
+    event.context = "Some Custom context"
+    return true
+}
+
+// run on single error only
+Bugsnag.notify(myThrowable) { event ->
+    event.context = "My Unique context"
+    return true
+}
+```
+
+If you use the NDK, you will need to implement your own native `on_error` callback _in addition_ to any JVM callbacks. This native `on_error` callback will run for fatal C/C++ errors only.
+
+See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.
+
+## Breadcrumb
+
+| v4.x API  | v5.x API |
+| ------------- | ------------- |
+| `Breadcrumb#getName`  | `Breadcrumb#getMessage` |
+
+## SessionTrackingPayload
+`SessionTrackingPayload` is now called `SessionPayload`.
+
+It is now possible to redact the `app` and `device` information captured on sessions via an `OnSessionCallback`.
+See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.
+
+## BeforeRecordBreadcrumb
+`BeforeRecordBreadcrumb` is now called `OnBreadcrumbCallback`.
+
+| v4.x API  | v5.x API |
+| ------------- | ------------- |
+| `BeforeRecordBreadcrumb#shouldRecord` | `OnBreadcrumbCallback#onBreadcrumb` |
+
+## Delivery
+
+| v4.x API  | v5.x API |
+| ------------- | ------------- |
+| `Delivery#deliver(Report, Configuration)` | `Delivery#deliver(SessionPayload, DeliveryParams)` |
+| `Delivery#deliver(SessionTrackingPayload, Configuration)` | `Delivery#deliver(Report, DeliveryParams)` |
```

Comment:
We need to pass the entire `Report` here rather than the `Event` as it contains all the information required for an Error Reporting API payload. I believe this will eventually be replaced by individual configuration options e.g. for setting a proxy